Implementing CounterStyleRegistry for reference resolution

vitorroriz opened the pull request:
#### 316bd081ae8deaba9d8933e555c173593b3a1600
<pre>
Implementing CounterStyleRegistry for reference resolution
<a href="https://bugs.webkit.org/show_bug.cgi?id=249870">https://bugs.webkit.org/show_bug.cgi?id=249870</a>
rdar://problem/103685490

Reviewed by Antti Koivisto and Myles C. Maxfield.

CSSCounterStyleRegistry keeps a collection of CounterStyle objects.
It is able to resolve references among the objects as needed by
the fallback feature and the extended system.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Adding CSSCounterStyleRegistry .cpp and .h files.

* Source/WebCore/css/CSSCounterStyleRegistry.cpp: Added.
(WebCore::CounterStyleRegistry::resolveReferences):
Resolve fallback and extend references for all the CounterStyle objects
in the registry.

(WebCore::CounterStyleRegistry::resolveExtendsReference):
Resolve Extend references for the counter styles with unresolved
references in the registry as described by
<a href="https://www.w3.org/TR/css-counter-styles-3/#extends-system.">https://www.w3.org/TR/css-counter-styles-3/#extends-system.</a>

(WebCore::CounterStyleRegistry::resolveFallbackReference):
Resolve fallback references for all counter styles with unresolved
references in the registry as described by
<a href="https://www.w3.org/TR/css-counter-styles-3/#counter-style-fallback.">https://www.w3.org/TR/css-counter-styles-3/#counter-style-fallback.</a>

(WebCore::CounterStyleRegistry::create):
(WebCore::CounterStyleRegistry::addCounterStyle):
Add counter style to the registry.

(WebCore::CounterStyleRegistry::getDecimalCounter):
(WebCore::CounterStyleRegistry::counterStyleForName):
Get counter style for a given name.

(WebCore::CounterStyleRegistry::userAgentCounterStyles):
(WebCore::CounterStyleRegistry::operator== const):

* Source/WebCore/css/CSSCounterStyleRegistry.h: Added.
* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSCounterStyleRule.h:
* Source/WebCore/css/CSSValueKeywords.in:
Adding CounterStyle as value to ListStyleType.

* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::effectiveListMarkerType):
(WebCore::listMarkerSuffix):
(WebCore::listMarkerText):
(WebCore::RenderListMarker::updateContent):
(WebCore::RenderListMarker::counterStyle const):
* Source/WebCore/rendering/RenderListMarker.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::counterStyleRegistry const):
(WebCore::RenderStyle::setCounterStyleRegistry):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialCounterStyleRegistry):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
Adding CounterStyle as value to ListStyleType enum.

* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
Adding a CounterStyleRegistry member to RenderStyle such that
the counter styles can be availble during rendering.

* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::isCacheable):
(): Deleted.

* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRules):
(WebCore::Style::RuleSetBuilder::addMutatingRulesToResolver):
Processing counter style rules and adding them as CounterStyle objects
to the CounterStyleRegistry hold by the StyleResolver.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueListStyleType):
(WebCore::Style::BuilderCustom::applyValueContent):
We can now handle custom-ident as list-style-type, as this is the
required type for describing a custom-style with an at-rule. For that
case, a CounterStyleRegistry will be recovered from the document&apos;s style
Scope and added to RenderStyle such that it can be used later for
rendering.  This branch will not be reached until custom-ident is added
as a type for list-style-type parser grammar in CSSProperties.json,
which will be defered until we can properly render the counter styles.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::ensureCounterStyleRegistry):
(WebCore::Style::Resolver::addCounterStyle):
(WebCore::Style::Resolver::counterStyleForName):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/StyleScope.h:
(WebCore::Style::Scope::resolverIfExists const):
(WebCore::Style::Scope::resolverIfExists): Deleted.
StyleResolver has now a CounterStyleRegistry for holding CounterStyle
objects added during the process of rules in RuleSetBuilder.
The registry will be recovered later when building the style for
list-style-type (StyleBuilderCustom.h)

Canonical link: <a href="https://commits.webkit.org/258510@main">https://commits.webkit.org/258510@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f73c19bcc6089b1e0470bf88964e42adfbc6fe15

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102150 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11295 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35221 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111475 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171645 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106132 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12274 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2208 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94531 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109210 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107929 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92674 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24155 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4846 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25579 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4962 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2017 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45076 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5834 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6707 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2022 Apple Inc.  All rights reserved.
```
Wrong year :)

litherum commented on Source/WebCore/dom/Document.h:
```diff
@@ -593,6 +594,9 @@ class Document
     const Style::CustomPropertyRegistry& customPropertyRegistry() const;
+    const CSSCounterStyleRegistry& counterStyleRegistry() const;
+    CSSCounterStyleRegistry& counterStyleRegistry();
+    // CSSCounterStyleRegistry* counterStyleRegistry();
```
We generally don't leave code commented. Please either delete it or uncomment it.

litherum commented on Source/WebCore/rendering/RenderListMarker.h:
```diff
@@ -67,7 +69,7 @@ class RenderListMarker final : public RenderBox {
     void styleDidChange(StyleDifference, const RenderStyle* oldStyle) final;
 
-    void element() const = delete;
+    // void element() const = delete;
```
Ditto.

litherum commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9020,6 +9020,16 @@ const Style::CustomPropertyRegistry& Document::customPropertyRegistry() const
 
+const CSSCounterStyleRegistry& Document::counterStyleRegistry() const
+{
+    return styleScope().counterStyleRegistry();
```
Is putting the registry in the `styleScope` the idiomatic way of doing this? For comparison, the document owns a `CSSFontSelector` directly. Maybe @anttijk knows?

Did you do this because it solves the lifetime problems of invalidating document state and rebuilding document state?

vitorroriz replied:
According to Antti, there was no StyleScope back in the time so things were being added directly to the document that got too extensive. I've moved it here as he did the same for CustomProperty.

anttijk replied:
Yea, it is fine. Style related things should generally go to style scope. Also many other things in Document should be factored out to reduce bloat.

Shadow scopes also potentially require their own registration though the CSSWG has not specced anything that requires it yet.

litherum commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -0,0 +1,156 @@
+
+void CSSCounterStyleRegistry::resolveReferences()
+{
+    // FIXME: does this needs thread-safety guards?
```
Why would it?

vitorroriz replied:
All elements will share this single registry now that we are not making one copy per RenderStyle, so I wanted to ask if we never render different elements concurrently.

litherum commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -0,0 +1,156 @@
+{
+    HashSet<CSSCounterStyle*> countersInChain;
+
+    std::function<void(CSSCounterStyle&)> resolve = [&](CSSCounterStyle& counterStyle) {
```
I don't understand why this is a lambda.

Also, `auto` is faster than `std::function` for lambdas.

vitorroriz replied:
It was auto before but the compiler can't deduce auto for recursive lambdas. I thought it was better than having a
 private member function that would only be used here, because it gets encapsulated, but I can change if you prefer it.

anttijk replied:
Let's just use static or member function if you need recursion until it is in the language (in C++23). std::function introduces an unnecessary virtual function (also we use WTF::Function).

litherum commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -0,0 +1,156 @@
+
+CounterStyleMap& CSSCounterStyleRegistry::userAgentCounterStyles()
+{
+    // FIXME: counter-style should get pre-defined counters from UA stylesheet rdar://103021161.
```
Should we dump the pre-defined ones into our user-agent style sheet instead of writing tons of C++ code to represent them?

vitorroriz replied:
Yes, this is definitely what we want to do. I've just added decimal manually because it is the last fallback option. Once we get other things in place we can add the counters to UA stylesheet including decimal.

litherum reviewed (approved):
I think this is a good first step, though I'd like @anttijk to comment on whether it makes sense for the styleScope to own the registry or not.

anttijk commented on Source/WebCore/css/CSSCounterStyleRegistry.h:
```diff
@@ -0,0 +1,58 @@
+    void resolveReferencesIfNeeded();
+    bool operator==(const CSSCounterStyleRegistry& other) const;
+    CSSCounterStyleRegistry() = default;
+private:
```
empty line before private:

anttijk commented on Source/WebCore/css/CSSCounterStyleRegistry.h:
```diff
@@ -0,0 +1,58 @@
+    bool operator==(const CSSCounterStyleRegistry& other) const;
+    CSSCounterStyleRegistry() = default;
+private:
+    using CounterStyleChain = HashSet<CSSCounterStyle*>;
```
You never use this type.

anttijk commented on Source/WebCore/css/CSSCounterStyleRegistry.h:
```diff
@@ -0,0 +1,58 @@
+    static CounterStyleMap& userAgentCounterStyles();
+    void resolveFallbackReference(CSSCounterStyle&);
+    void resolveExtendsReference(CSSCounterStyle&);
+    RefPtr<CSSCounterStyle> getCounterStyle(const AtomString&);
```
"get" naming

anttijk commented on Source/WebCore/css/CSSCounterStyleRegistry.h:
```diff
@@ -0,0 +1,58 @@
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    RefPtr<CSSCounterStyle> getResolvedCounterStyle(const AtomString&);
+    RefPtr<CSSCounterStyle> getDecimalCounter() const;
```
Coding style tells to not start function name with "get". You can just drop the word in these two, it adds nothing.

anttijk commented on Source/WebCore/style/RuleSetBuilder.cpp:
```diff
@@ -377,11 +378,18 @@ void RuleSetBuilder::addMutatingRulesToResolver()
             // FIXME: Use the shadow tree scope if applicable (or just skip).
             auto& registry = m_resolver->document().styleScope().customPropertyRegistry();
             registry.registerFromStylesheet(downcast<StyleRuleProperty>(rule.get()).descriptor());
-            continue;
```
`continue` is here for symmetry, and so next person adding a rule type doesn't forget it. No need to drop it.

vitorroriz committed:
Implementing CounterStyleRegistry for reference resolution
https://bugs.webkit.org/show_bug.cgi?id=249870
rdar://problem/103685490

Reviewed by Antti Koivisto and Myles C. Maxfield.

CSSCounterStyleRegistry keeps a collection of CounterStyle objects.
It is able to resolve references among the objects as needed by
the fallback feature and the extended system.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Adding CSSCounterStyleRegistry .cpp and .h files.

* Source/WebCore/css/CSSCounterStyleRegistry.cpp: Added.
(WebCore::CounterStyleRegistry::resolveReferences):
Resolve fallback and extend references for all the CounterStyle objects
in the registry.

(WebCore::CounterStyleRegistry::resolveExtendsReference):
Resolve Extend references for the counter styles with unresolved
references in the registry as described by
https://www.w3.org/TR/css-counter-styles-3/#extends-system.

(WebCore::CounterStyleRegistry::resolveFallbackReference):
Resolve fallback references for all counter styles with unresolved
references in the registry as described by
https://www.w3.org/TR/css-counter-styles-3/#counter-style-fallback.

(WebCore::CounterStyleRegistry::create):
(WebCore::CounterStyleRegistry::addCounterStyle):
Add counter style to the registry.

(WebCore::CounterStyleRegistry::getDecimalCounter):
(WebCore::CounterStyleRegistry::counterStyleForName):
Get counter style for a given name.

(WebCore::CounterStyleRegistry::userAgentCounterStyles):
(WebCore::CounterStyleRegistry::operator== const):

* Source/WebCore/css/CSSCounterStyleRegistry.h: Added.
* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSCounterStyleRule.h:
* Source/WebCore/css/CSSValueKeywords.in:
Adding CounterStyle as value to ListStyleType.

* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::effectiveListMarkerType):
(WebCore::listMarkerSuffix):
(WebCore::listMarkerText):
(WebCore::RenderListMarker::updateContent):
(WebCore::RenderListMarker::counterStyle const):
* Source/WebCore/rendering/RenderListMarker.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::counterStyleRegistry const):
(WebCore::RenderStyle::setCounterStyleRegistry):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialCounterStyleRegistry):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
Adding CounterStyle as value to ListStyleType enum.

* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
Adding a CounterStyleRegistry member to RenderStyle such that
the counter styles can be availble during rendering.

* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::isCacheable):
(): Deleted.

* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRules):
(WebCore::Style::RuleSetBuilder::addMutatingRulesToResolver):
Processing counter style rules and adding them as CounterStyle objects
to the CounterStyleRegistry hold by the StyleResolver.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueListStyleType):
(WebCore::Style::BuilderCustom::applyValueContent):
We can now handle custom-ident as list-style-type, as this is the
required type for describing a custom-style with an at-rule. For that
case, a CounterStyleRegistry will be recovered from the document's style
Scope and added to RenderStyle such that it can be used later for
rendering.  This branch will not be reached until custom-ident is added
as a type for list-style-type parser grammar in CSSProperties.json,
which will be defered until we can properly render the counter styles.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::ensureCounterStyleRegistry):
(WebCore::Style::Resolver::addCounterStyle):
(WebCore::Style::Resolver::counterStyleForName):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/StyleScope.h:
(WebCore::Style::Scope::resolverIfExists const):
(WebCore::Style::Scope::resolverIfExists): Deleted.
StyleResolver has now a CounterStyleRegistry for holding CounterStyle
objects added during the process of rules in RuleSetBuilder.
The registry will be recovered later when building the style for
list-style-type (StyleBuilderCustom.h)

Canonical link: https://commits.webkit.org/258510@main
